feat: add evalTaming option (#961)
endojs/endo: packages/ses/docs/lockdown.md

However, in the shim, only the exact `eval` function from the start compartment can be used to
perform direct-eval, which runs in the lexical scope in which the direct-eval syntax appears (direct-eval is a special form rather than a function call).
The SES shim itself uses direct-eval internally to construct an isolated
evaluator, so replacing the initial `eval` prevents any subsequent program
from using the same mechanism to isolate a guest program.

The `"unsafeEval"` option for `evalTaming` leaves the original `eval` in place
for other isolation mechanisms like isolation code generators that work in
tandem with SES.
This option may be useful for web pages with a Content Security Policy
that excludes `unsafe-eval` or browser extension environments with similar
restrictions, or development-mode bundling systems that use `eval`.
In these cases, SES cannot be responsible for maintaining the isolation of
guest code. If you're going to use `eval`, [Trusted
Types](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/trusted-types) may help maintain security.

> The `"noEval"` option emulates a Content Security Policy that excludes

Author: ```suggestion
The `"noEval"` option emulates a Content Security Policy that disallows
```

- `'none'`: do not install traps for uncaught exceptions. Errors are likely to
  appear as `{}` when they are reported by the default trap.

## `evalTaming` Options

> This option only affects the initial compartment!

Reviewer: I've been using "start compartment" rather than "initial compartment", but I defer to @kriskowal on this.

```suggestion
This option only affects the start compartment!
```

> **Background**: For every compartment including the implied compartment in the
> initial JavaScript realm, there are evaluators `eval` and `Function`.

Reviewer: Whether the realm is the initial realm doesn't matter. The description here applies to the compartments of any realm.

```suggestion
**Background**: Every realm has an implicit initial compartment we call the "start compartment". Explicit compartments are made with the `Compartment` constructor.
For every compartment including the start compartment, there are evaluators `eval` and `Function`.
```

This does seem like a weird place to define "start compartment" and "explicit compartment". If they are adequately explained elsewhere, we can delete my first sentence above.

The default lockdown behavior isolates all of these evaluators.

Replacing the realm's initial evaluators is not necessary to ensure the

> isolation of guest code because guest code must not run in the initial realm.

Reviewer: ```suggestion
isolation of guest code because guest code must not run in the start compartment.
```

> However, replacing these evaluators is the safest and therefore default
> evaluator taming option (`"safeEval"`) because it may limit the harm that guest code can cause if it escapes its assigned compartment.

Reviewer: "escapes its assigned compartment" is not the main threat. Perhaps something like the following?

```suggestion
Although the code run in the start compartment is normally referred to as "trusted", we mean only that we assume it was not written maliciously. It may still be buggy, and it may be buggy in a way that is exploitable by malicious guest code. To limit the hard that such vulnerabilities can cause, the default (`"safeEval"`) setting replaces the evaluators of the start compartment with their safe alternatives.
```

> However, only the exact `eval` function from the initial realm can be used to

Reviewer: ```suggestion
However, in the shim, only the exact `eval` function from the start compartment can be used to
```
I added "in the shim" because the hardened JS proposal and (IIUC) the XS implementation have no such restriction on direct eval.

perform direct-eval, which runs in the lexical scope in which the direct-eval syntax appears (direct-eval is a special form rather than a function call).
The SES shim itself uses direct-eval internally to construct an isolated
evaluator, so replacing the initial `eval` prevents any subsequent program
from using the same mechanism to isolate a guest program.

> The `"unsafe"` option for `evalTaming` leaves the original `eval` in place

Reviewer 2: ```suggestion
The `"unsafeEval"` option for `evalTaming` leaves the original `eval` in place
```

for other isolation mechanisms like isolation code generators that work in
tandem with SES.
This option may be useful for web pages with a Content Security Policy

> that excludes `unsafe-eval` or browser extension environments with similar

Reviewer: This seems contradicted by
> ```js
> lockdown({ evalTaming: 'unsafeEval' });
> // !!!! DO NOT USE IT IN PRODUCTION WITH "unsafe-eval" in CSP or even no CSP !!!!
> ```

below

Author: > web pages with a Content Security Policy that excludes `unsafe-eval` or browser extension environments

I think it should be "includes" other than "excludes"

Author: or "allows `unsafe-eval`"

restrictions, or development-mode bundling systems that use `eval`.
In these cases, SES cannot be responsible for maintaining the isolation of
guest code. If you're going to use `eval`, [Trusted
Types](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/trusted-types) may help maintain security.

The `"noEval"` option emulates a Content Security Policy that excludes
`unsafe-eval` by replacing all evaluators with functions that throw an

> exception.

Reviewer 3: Shouldn't `compartment.evaluate()` also throw in this case?

Reviewer: I think so. But if so, then the first sentence above
> This option only affects the initial compartment!

needs revision.

Reviewer 3: Actually thinking about this more, since `compartment.evaluate()` evaluates in the context of the child compartment, `noEval` as implemented currently still only applies to the start compartment, so it is consistent.

This section of the doc should be updated to highlight that only evaluators for the start compartment are neutered, but that child compartments remain with safe evaluators. Those can be tamed explicitly by the program if needed (by just copying the throwing noEval `eval` and `Function` from the start compartment into the global object of the child compartment)

Author: > Those can be tamed explicitly by the program if needed (by just copying the throwing noEval eval and Function from the start compartment into the global object of the child compartment)

Or also adding it to the Compartment constructor? Therefore devs can get the noEval in the Compartment without make the initial compartment also using noEval

```js
const c = new Compartment({}, { evalTaming: 'noEval' })
c.evaluate('1+1') // ok
c.globalThis.noEval('1+1') // no
```

Reviewer 2: I can’t think of a reason that would motivate this level of specificity.

```js
lockdown(); // evalTaming defaults to 'safeEval'
// or
lockdown({ evalTaming: 'noEval' }); // disallowing calling eval like there is a CSP limitation.
// or
lockdown({ evalTaming: 'unsafeEval' });
// !!!! DO NOT USE IT IN PRODUCTION WITH "unsafe-eval" in CSP or even no CSP !!!!
```

## `stackFiltering` Options

**Background**: The error stacks shown by many JavaScript engines are
voluminous.
They contain many stack frames of functions in the infrastructure, that is

> usually irrelevant to the programmer trying to diagnose a bug. The SES-shim's

Reviewer 2: 🙏

- `'none'`: do not install traps for uncaught exceptions. Errors are likely to
  appear as `{}` when they are reported by the default trap.

## `evalTaming` Options

This option only affects the start compartment!

To disallows eval in the explicit compartments, replace the constructors in the compartment.

```js
const c = new Compartment()

> c.globalThis.eval = c.globalThis.Function = function() {

Author: wait, what about hidden intrinsics? 🤔

AsyncFunction, GeneratorFunction, ...?

Author: For a compartment, its own evaluator is not reachable unless the reference to itself is leaked.

Does it mean I should do some change on this PR?

- `'none'`: do not install traps for uncaught exceptions. Errors are likely to
  appear as `{}` when they are reported by the default trap.

## `evalTaming` Options

> **Background**: In some environment, development mode allows `unsafe-eval` but the production mode does not.
> This option allows disabling the replacement of `eval` and `Function` so that things like `eval-source-map` works in development mode.
> Assume no `eval-source-map` (or things like that) is used in production, and no `eval` is allowed (e.g. by Content Security Policy) in production, taming `eval` and `Function` constructor actally hurts developer experience.
> 
> This option should ONLY be used when you have situation like described above or you think you have reasonable use case to do so.
> Security audit is NOT performed on this option and you SHOULD NEVER use it in production.
> 
> Strongly suggest to use [Trusted Types](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/trusted-types) when you have use cases like this.
> 
> This option only applies to the start compartment. Child compartment will always use value `"safeEval"`.

Reviewer 2: ```suggestion
**Background**: Every compartment including the implied compartment in the
initial JavaScript realm, there are evaluators `eval` and `Function`.
The default lockdown behavior isolates all of these evaluators.

Replacing the realm's initial evaluators is not necessary to ensure the
isolation of guest code because guest code must not run in the initial realm.
However, replacing these evaluators is the safest and therefore default
evaluator taming option (`"safe"`) because it may limit the harm that guest code can cause
if it escapes its assigned compartment.

However, only the exact `eval` function from the initial realm can be used to
perform direct-eval, which runs in the caller's scope (dynamic scope).
The SES shim itself uses direct-eval internally to construct an isolated
evaluator, so replacing the initial `eval` prevents any subsequent program
from using the same mechanism to isolate a guest program.

The `"unsafe"` option for `evalTaming` leaves the original `eval` in place
for other isolation mechanisms like isolation code generators that work in
tandem with SES.
This option may be useful in may for web pages with a Content Security Policy
that excludes `unsafe-eval` or browser extension environments with similar
restrictions, or development-mode bundling systems that use `eval`.
In these cases, SES cannot be responsible for maintaining the isolation of
guest code.
Using [Trusted
Types](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/trusted-types)
in combination with `"unsafe"` `evalTaming` may help maintain isolation.

The `"none"` option emulates a Content Security Policy that excludes
`unsafe-eval` by replacing all evaluators with functions that throw an
exception, in the initial realm as well as any derived compartment.
```

## `evalTaming` Options

> **Background**: Every compartment including the implied compartment in the

Reviewer: ```suggestion
**Background**: In every compartment including the implied compartment in the
```
or
```suggestion
**Background**: For every compartment including the implied compartment in the
```

initial JavaScript realm, there are evaluators `eval` and `Function`.
The default lockdown behavior isolates all of these evaluators.

Replacing the realm's initial evaluators is not necessary to ensure the
isolation of guest code because guest code must not run in the initial realm.
However, replacing these evaluators is the safest and therefore default

> evaluator taming option (`"safe"`) because it may limit the harm that guest code can cause

Reviewer: Each time one of these is mentioned, please use the exact literal string needed to avoid confusing the reader.

```suggestion
evaluator taming option (`"safeEval"`) because it may limit the harm that guest code can cause
```

if it escapes its assigned compartment.

However, only the exact `eval` function from the initial realm can be used to

> perform direct-eval, which runs in the caller's scope (dynamic scope).

Reviewer: ```suggestion
perform direct-eval, which runs in the lexical scope in which the direct-eval syntax appears (direct-eval is a special form rather than a function call).
```
I can see the rationale for calling this dynamic scope, but I think it is better to think of direct-eval as a special form, like `if` is, and therefore naturally executes in the lexical scope of the code it appears in. Unfortunately, I doubt the term "special form" is well known outside Scheme. Better terminology would be good, but go with "special form" if we don't come up with anything better.

Or, frankly, you could just omit the parenthetical. It's ok if this document isn't the one to try to explain anything more about direct-eval.
```suggestion
perform direct-eval, which runs in the lexical scope in which the direct-eval syntax appears.
```

The SES shim itself uses direct-eval internally to construct an isolated
evaluator, so replacing the initial `eval` prevents any subsequent program
from using the same mechanism to isolate a guest program.

The `"unsafe"` option for `evalTaming` leaves the original `eval` in place
for other isolation mechanisms like isolation code generators that work in
tandem with SES.

> This option may be useful in may for web pages with a Content Security Policy

Reviewer: ```suggestion
This option may be useful for web pages with a Content Security Policy
```

that excludes `unsafe-eval` or browser extension environments with similar
restrictions, or development-mode bundling systems that use `eval`.
In these cases, SES cannot be responsible for maintaining the isolation of
guest code.
Using [Trusted
Types](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/trusted-types)

> in combination with `"unsafe"` `evalTaming` may help maintain isolation.

Reviewer: Which "unsafe" do you mean here?

> The `"none"` option emulates a Content Security Policy that excludes

Reviewer: ```suggestion
The `"noEval"` option emulates a Content Security Policy that excludes
```

## `evalTaming` Options

This option only affects the start compartment!

**Background**: Every realm has an implicit initial compartment we call the "start compartment". Explicit compartments are made with the `Compartment` constructor.
For every compartment including the start compartment, there are evaluators `eval` and `Function`.
The default lockdown behavior isolates all of these evaluators.

Replacing the realm's initial evaluators is not necessary to ensure the
isolation of guest code because guest code must not run in the start compartment.

> Although the code run in the start compartment is normally referred to as "trusted", we mean only that we assume it was not written maliciously. It may still be buggy, and it may be buggy in a way that is exploitable by malicious guest code. To limit the hard that such vulnerabilities can cause, the default (`"safeEval"`) setting replaces the evaluators of the start compartment with their safe alternatives.

Reviewer: ```suggestion
Although the code run in the start compartment is normally referred to as "trusted", we mean only that we assume it was not written maliciously. It may still be buggy, and it may be buggy in a way that is exploitable by malicious guest code. To limit the harm that such vulnerabilities can cause, the default (`"safeEval"`) setting replaces the evaluators of the start compartment with their safe alternatives.
```
